### PR TITLE
LibCore: Introduce and use Core::Environment wrapper for FOOenv() functions

### PR DESCRIPTION
**Pull request by AtkinsSJ**
Another LibC thing that I felt deserved a nicer wrapper. A couple of these we had wrappers for in Core::System, but not all of them, and the environment is a separate enough thing to belong on its own. So, the existing wrapper functions are moved to `Core::Environment`, and the others filled in.

Initially I thought of keeping a HashMap around of the environment variables, but given that the environ can be modified from other places, it made more sense to just wrap the LibC functions and provide a nicer API.

Of course, "nicer" is my opinion, so I could be wrong. Feedback welcome.

### CODE REVIEW

**Comment by AtkinsSJ** [comment]
CI failure is legitimate, I need to put replacement implementations in for `clearenv()` and `secure_getenv()` on MacOS.

File: Userland/Libraries/LibCore/Environment.cpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    return Entry {
+        .full_entry = input,
+        .name = input.substring_view(0, *split_index),
+        .value = input.substring_view(*split_index + 1),
+    };
+}
+
+void for_each_entry(Function<IterationDecision(Entry&)> callback)
```
Comment (DanShaders) [inline Userland/Libraries/LibCore/Environment.cpp:49]: Not sure how idiomatic it is but range-based-for-loop-iteratable `Core::Environment::entries()` would be really cool. (I really don't like `Function`s where not necessary)

File: Userland/Libraries/LibCore/Environment.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        entries.append(entry);
+        return IterationDecision::Continue;
+    });
+
+    return entries;
+}
+
+Optional<StringView> get(StringView name)
+{
+    auto* result = ::getenv(name.to_byte_string().characters());
```
Comment (DanShaders) [inline Userland/Libraries/LibCore/Environment.cpp:84]: Just do `for_each_entry` and there won't be a ByteString allocation.

File: Userland/Libraries/LibCore/Environment.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    if (result)
+        return StringView { result, strlen(result) };
+    return {};
+}
+
+Optional<StringView> get_secure(StringView name)
+{
+    auto* result = ::secure_getenv(name.to_byte_string().characters());
```
Comment (DanShaders) [inline Userland/Libraries/LibCore/Environment.cpp:92]: Same here, though a call to `getauxval(AT_SECURE)` will be required.

File: Userland/Libraries/LibCore/Environment.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    TRY(builder.try_append('\0'));
+
+    // Note the explicit null terminator above.
+    auto rc = ::unsetenv(builder.string_view().characters_without_null_termination());
+    if (rc < 0)
+        return Error::from_errno(errno);
+    return {};
+}
+
+ErrorOr<void> put(StringView env)
```
Comment (DanShaders) [inline Userland/Libraries/LibCore/Environment.cpp:127]: Is there any reason to prefer this flawed `putenv` over `setenv`?

**Review by DanShaders — COMMENTED** [review]
I decided I might chime in since I designed a similar API for an unmerged LibRuntime.

File: Userland/Libraries/LibCore/Environment.cpp (hunk 1 above)
Comment (alimpfard) [reply Userland/Libraries/LibCore/Environment.cpp:49]: +1 for begin()/end() instead, this is basically just a simple index range plus a mapper

File: Userland/Libraries/LibCore/Environment.cpp [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+    return environ_size;
+}
+
+Vector<Entry> entries()
+{
+    Vector<Entry> entries;
+
+    for_each_entry([&entries](Entry& entry) {
+        entries.append(entry);
```
Comment (alimpfard) [inline Userland/Libraries/LibCore/Environment.cpp:75]: This feels...odd, I'd prefer being given ownership of the entry, the for_each function throws it away after passing it to the callback function anyway.

File: Userland/Libraries/LibCore/Environment.cpp (hunk 3 above)
Comment (alimpfard) [reply Userland/Libraries/LibCore/Environment.cpp:92]: Same StringBuilder trick applies here.

File: Userland/Libraries/LibCore/Environment.cpp [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+ErrorOr<void> set(StringView name, StringView value, Overwrite overwrite)
+{
+    auto builder = TRY(StringBuilder::create());
+    TRY(builder.try_append(name));
+    TRY(builder.try_append('\0'));
+    TRY(builder.try_append(value));
+    TRY(builder.try_append('\0'));
+    // Note the explicit null terminators above.
+    auto c_name = builder.string_view().characters_without_null_termination();
```
Comment (alimpfard) [inline Userland/Libraries/LibCore/Environment.cpp:106]: Lovely, see, you've done it here already :P

File: Userland/Utilities/run-tests.cpp [hunk 7]
```diff
@@ -360,15 +361,15 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
 
     test_glob = ByteString::formatted("*{}*", test_glob);
 
-    if (getenv("DISABLE_DBG_OUTPUT")) {
+    if (Core::Environment::get("DISABLE_DBG_OUTPUT"sv).has_value()) {
```
Comment (alimpfard) [inline Userland/Utilities/run-tests.cpp:364]: Might be worth considering a `C::E::has(name)` API?

File: Ladybird/Utilities.cpp [hunk 8]
```diff
@@ -24,13 +25,14 @@ ErrorOr<String> application_directory()
...
+        if (auto source_dir = Core::Environment::get("SERENITY_SOURCE_DIR"sv); source_dir.has_value()) {
+            return ByteString::formatted("{}/Base", *source_dir);
         }
-        auto* home = getenv("XDG_CONFIG_HOME") ?: getenv("HOME");
-        VERIFY(home);
-        auto home_lagom = ByteString::formatted("{}/.lagom", home);
+        auto home = Core::Environment::get("XDG_CONFIG_HOME"sv);
+        if (!home.has_value())
+            home = Core::Environment::get("HOME"sv);
+        VERIFY(home.has_value());
```
Comment (alimpfard) [inline Ladybird/Utilities.cpp:34]: This verify is already present in `*home`, though I don't really mind having it here too.

File: Userland/Libraries/LibCore/System.cpp [hunk 9]
```diff
@@ -1273,19 +1274,14 @@ ErrorOr<void> adjtime(const struct timeval* delta, struct timeval* old_delta)
...
 {
     Vector<StringView> exec_environment;
-    for (size_t i = 0; environ[i]; ++i) {
-        StringView env_view { environ[i], strlen(environ[i]) };
-        auto maybe_needle = env_view.find('=');
-
-        if (!maybe_needle.has_value())
-            continue;
-
+    Environment::for_each_entry([&](Environment::Entry& entry) {
```
Comment (alimpfard) [inline Userland/Libraries/LibCore/System.cpp:1277]: Yeah I really don't like how you can't tell what you're allowed to do with this Entry.
One might expect that modifying this object would modify the environment too, not a good thing.

File: Userland/Shell/Builtin.cpp [hunk 10]
```diff
@@ -636,8 +637,10 @@ ErrorOr<int> Shell::builtin_export(Main::Arguments arguments)
         return 1;
 
     if (vars.is_empty()) {
-        for (size_t i = 0; environ[i]; ++i)
-            puts(environ[i]);
+        Core::Environment::for_each_entry([](Core::Environment::Entry& entry) {
+            outln(entry.full_entry);
```
Comment (alimpfard) [inline Userland/Shell/Builtin.cpp:641]: `outln("{}", entry.full_entry)`

**Review by alimpfard — COMMENTED** [review]
API is pretty nice, but could use some improvements (plus a couple nits)

File: Userland/Libraries/LibCore/Environment.cpp [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+ * Copyright (c) 2021-2024, Sam Atkins <atkinssj@serenityos.org>
+ * Copyright (c) 2022, Matthias Zimmerman <matthias291999@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Environment.h"
+#include <AK/ByteString.h>
+
+#if defined(AK_OS_MACOS)
```
Comment (shannonbooth) [inline Userland/Libraries/LibCore/Environment.cpp:13]: There's an equivalent include in `System.cpp` which can removed in lieu of this one now I believe

File: Userland/Libraries/LibCore/Environment.cpp (hunk 4 above)
Comment (AtkinsSJ) [reply Userland/Libraries/LibCore/Environment.cpp:127]: :man_shrugging: I'm just matching the LibC API.

**Comment by AtkinsSJ** [comment]
**Changes:**
- Rebase!
- Fix MacOS (hopefully???): Fall back to regular getenv(), and copy the implementation of clearenv() from LibC.
- Replace the goofy `for_each_entry()` with an iterator. I know very little about iterators so I might have made a boo-boo so look out for that.
- Use StringBuilders instead of `to_byte_string()`. This technically can produce OOM errors, and the other places TRY() those because they already return ErrorOr, but having getenv() be fallible feels weird so I just made it use the non-errory functions.
- Added `Core::Environment::has(name)` for that one case. It's pretty nice, I won't lie.
- Correct that `outln()` usage.
- Remove that now-unneeded MacOS include from System.cpp

@alimpfard does this fix your ownership concerns? I'm not 100% sure I understood that.

File: Userland/Libraries/LibCore/Environment.cpp (hunk 4 above)
Comment (DanShaders) [reply Userland/Libraries/LibCore/Environment.cpp:144]: You definitely don't have to, even POSIX says, "The setenv() function is preferred over this function."

The single use in SS/Service.cpp can be safely changed to setenv.

**Comment by DanShaders** [comment]
> Use StringBuilders instead of to_byte_string().

And
```cpp
for (auto entry : entries())
  if (entry.name == name)
    return entry.value;
return {};
```
would be both infallible and shorter...

File: Userland/Libraries/LibCore/Environment.h [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+        return *this;
+    }
+    constexpr EntryIterator operator++(int)
+    {
+        auto result = *this;
+        ++m_index;
+        return result;
+    }
+
+    Entry const operator*() const { return Entry::from_chars(raw_environ()[m_index]); }
```
Comment (alimpfard) [inline Userland/Libraries/LibCore/Environment.h:46]: Don't really need the const, but yes, this (non-reference return) addresses my ownership problem.

**Comment by AtkinsSJ** [comment]
**Changes:**
- Iterate `entries()` instead of calling `getenv()`
- Replace `get_secure()` with a second parameter to `Environment::get()`, which defaults to non-secure.
- Remove unnecessary const qualifiers in the Entry iterator.
- Rebase on master to make sure nothing's broken. (Second push)

**Comment by AtkinsSJ** [comment]
**Changes:**
- Go back to calling `getenv()`/`secure_getenv()` directly, because otherwise we have to include `<sys/auxv.h>` which doesn't always exist, including on CI!
- Rename `Environment::Secure` to `Environment::SecureOnly`.

**Comment by ADKaster** [comment]
Missing a header for `environ` on macOS

`/Users/runner/work/1/s/Userland/Libraries/LibCore/Environment.cpp:61:24: error: use of undeclared identifier 'environ'`

https://dev.azure.com/SerenityOS/SerenityOS/_build/results?buildId=44741&view=logs&j=2351e412-c793-598f-27f8-72c4f2d55ba1&t=9c50333a-ada2-5596-d44b-d8dcee270ad2&l=99

**Comment by AtkinsSJ** [comment]
> Missing a header for `environ` on macOS
> 
> `/Users/runner/work/1/s/Userland/Libraries/LibCore/Environment.cpp:61:24: error: use of undeclared identifier 'environ'`
> 
> [dev.azure.com/SerenityOS/SerenityOS/_build/results?buildId=44741&view=logs&j=2351e412-c793-598f-27f8-72c4f2d55ba1&t=9c50333a-ada2-5596-d44b-d8dcee270ad2&l=99](https://dev.azure.com/SerenityOS/SerenityOS/_build/results?buildId=44741&view=logs&j=2351e412-c793-598f-27f8-72c4f2d55ba1&t=9c50333a-ada2-5596-d44b-d8dcee270ad2&l=99)

**Changes:**
- Use the `environ` wrapper so things work on macOS. I'm... really surprised this didn't come up before, I was using `environ` in multiple places.

**Comment by AtkinsSJ** [comment]
**Changes:**
- Another macOS fix, marking a parameter as `[[maybe_unused]]` which isn't used in the macOS code. Deary me, macOS will be the end of me. :laughing:

**Comment by AtkinsSJ** [comment]
**Changes:**
- Fixed the conflict. :smiling_face_with_tear:

**Review by ADKaster — APPROVED** [review]

**Pull request merged**